### PR TITLE
Support a sharing lru cache.

### LINKED ISSUE #228: Shared cache?
Looks like GetBlockCacher isn't called anywhere?
In fact it looks like the only thing used from the options passed to *DB is the cache capacity https://github.com/syndtr/goleveldb/blob/master/leveldb/table.go#L454 which it uses and directly creates an LRU.

I'm interested in passing in a cache.NewLRU to more than one database that is larger than it could be if each had their own. Because of the namespace that the cache supports already and seems to use file descriptor number, it should be safe correct?

### PR DESCRIPTION
**Pull request by yonderblue**
Fixes #228

Should be backwards compatible, and if this approach is good perhaps some tests are appropriate.
Example main() pseudo usage:
```
main () {
	sharedLRU := cache.NewSharedLRU(opt.MiB * 16)

	var namespaceInc byte

	getCacher := func() opt.Cacher {
		ns := namespaceInc
		namespaceInc++

		return &opt.CacherFunc{
			NewFunc: func(int) cache.Cacher {
				return cache.NewNamespaceLRU(ns, sharedLRU)
			},
		}
	}
	
	d1, err := leveldb.OpenFile(p1, &opt.Options{BlockCacher: getCacher()})
	...
	
	d2, err := leveldb.OpenFile(p2, &opt.Options{BlockCacher: getCacher()})
	...
	
	d3, err := leveldb.OpenFile(p2, &opt.Options{BlockCacher: getCacher()})
	...
}
```

### CODE REVIEW

**Comment by gaillard** [comment]
Any concerns here?

File: .travis.yml [hunk 1]
```diff
@@ -1,9 +1,6 @@
 language: go
```
Comment (tgulacsi) [inline .travis.yml:1]: Please do not change .travis.yml in thisn commit!

File: leveldb/cache/lru.go [hunk 2]
```diff
@@ -169,27 +172,57 @@ func (r *lru) EvictNS(ns uint64) {
...
-func NewLRU(capacity int) Cacher {
-	r := &lru{capacity: capacity}
+// NewSharedLRU creates a new shared lru.
+func NewSharedLRU(capacity int) *SharedLRU {
+	r := &SharedLRU{capacity: capacity}
 	r.reset()
 	return r
 }
+
+// NewLRU create a new namespaced LRU-cache.
```
Comment (tgulacsi) [inline leveldb/cache/lru.go:219]: The comment should start with the function's name.

File: leveldb/cache/lru.go [hunk 3]
```diff
@@ -148,19 +149,21 @@ func (r *lru) Evict(n *Node) {
 	rn.h.Release()
 }
 
-func (r *lru) EvictNS(ns uint64) {
+//nodeNS optional.
```
Comment (tgulacsi) [inline leveldb/cache/lru.go:152]:
```
// evictMulti evicts multiple nodes
// nodeNS is optional
```

File: leveldb/cache/lru.go [hunk 4]
```diff
@@ -169,27 +172,57 @@ func (r *lru) EvictNS(ns uint64) {
...
-func (r *lru) EvictAll() {
-	r.mu.Lock()
-	back := r.recent.prev
-	for rn := back; rn != &r.recent; rn = rn.prev {
-		rn.n.CacheData = nil
-	}
-	r.reset()
-	r.mu.Unlock()
+type lru struct {
+	lruNS byte
```
Comment (tgulacsi) [inline leveldb/cache/lru.go:176]: Change the order to `*SharedLRU,byte` to save on alignment.

**Review by tgulacsi — CHANGES_REQUESTED** [review]

**Comment by syndtr** [comment]
Superseded by 4bbca4ea2470b1c244f502710f0cc52023ba0e44.

**Pull request closed without merging**